### PR TITLE
Fixup for smithy generated clients imports

### PR DESCRIPTION
**Pull request by jasdel**
Updates the import paths used by generated smithy API clients. Also minor tweaks to generated code to compile.

### CODE REVIEW

File: io/ringbuffer.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package sdkio
```
Comment (skotambkar) [inline io/ringbuffer.go:1]: should the package name still be sdkio?

File: transport/http/client.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	return ClientHandler{
+		client: client,
+	}
+}
+
+// Handle implements the middleware Handler interface, that will invoke the
+// underlying HTTP client. Requires the input to be an Smithy *Request. Returns
+// a smithy *Response, or error if the request failed.
+func (c ClientHandler) Handle(ctx context.Context, input interface{}) (
+	out interface{}, metadata middleware.Metadata, err error,
```
Comment (skotambkar) [inline transport/http/client.go:42]: out is not used anywhere; out will have rawresponse, so may be we should expose that somehow?

File: errors.go [hunk 3]
```diff
@@ -60,6 +60,21 @@ func (e *OperationError) Error() string {
...
+// deserialization.
+type DeserializationError struct {
+	Err      error //  original error
+	Snapshot []byte
+}
+
+// Error returns a formatted error for DeserializationError
+func (e *DeserializationError) Error() string {
+	return fmt.Sprintf("deserialization failed, %v", e.Err)
+}
```
Comment (skotambkar) [inline errors.go:73]: Nit: I have the change in my improvement PR, basically this should check if e.Err is nil. Similar to how SerializationError is doing. 

```
// Error returns a formatted error for DeserializationError
func (e *DeserializationError) Error() string {
	const msg = "deserialization failed"
	if e.Err == nil {
		return msg
	}
	return fmt.Sprintf("%s: %v", msg, e.Err)
}

```

File: io/byte.go [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package sdkio
```
Comment (skmcgrail) [inline io/byte.go:1]: package name does not match path

File: io/ringbuffer.go (hunk 1 above)
Comment (skmcgrail) [inline io/ringbuffer.go:1]: package name should be `io`

File: io/ringbuffer_test.go [hunk 5]
```diff
@@ -0,0 +1,320 @@
+package sdkio
```
Comment (skmcgrail) [inline io/ringbuffer_test.go:1]: package name `io`

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java [hunk 6]
```diff
@@ -54,6 +54,8 @@
             "github.com/awslabs/smithy-go/httpbinding", null, Versions.SMITHY_GO),
     SMITHY_JSON("dependency", "github.com/awslabs/smithy-go",
             "github.com/awslabs/smithy-go/json", "smithyjson", Versions.SMITHY_GO),
+    SMITHY_IO("dependency", "github.com/awslabs/smithy-go",
```
Comment (skotambkar) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java:57]: I don't see this being used, but usage might be in another PR. Just want to call out

File: io/byte.go (hunk 4 above)
Comment (skotambkar) [reply io/byte.go:1]: Ah had the same comment

File: transport/http/client.go (hunk 2 above)
Comment (jasdel) [reply transport/http/client.go:42]: Still needed for the middleware.Handler interface though. The deserialization middleware is where this `out` value will be used as the RawResponse

File: errors.go (hunk 3 above)
Comment (jasdel) [reply errors.go:73]: cool will update.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java (hunk 6 above)
Comment (jasdel) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoDependency.java:57]: This is used by the aws-sdk-go-v2 pr that i've not posted yet.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java [hunk 7]
```diff
@@ -975,34 +980,35 @@ private void writeRestDeserializerMember(
...
-                            CodegenUtils.generatePointerReferenceIfPointable(targetShape, value));
+                            CodegenUtils.generatePointerValueIfPointable(writer, targetShape, value));
                 });
     }
 
     private void writePrefixHeaderDeserializerFunction(
-            GoWriter writer, Model model, String memberName,
-            Shape targetShape, HttpBinding binding
+            GoWriter writer, Model model, SymbolProvider symbolProvider,
+            String memberName, Shape targetShape, HttpBinding binding
```
Comment (skmcgrail) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java:1011]: nit: Style says that for function parameter definitions that span lines should be formatted such that each parameter is on its own line.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java [hunk 8]
```diff
@@ -975,34 +980,35 @@ private void writeRestDeserializerMember(
...
             default:
                 throw new CodegenException("unexpected http binding found");
         }
     }
 
     private void writeHeaderDeserializerFunction(
-            GoWriter writer, Model model, String memberName,
-            Shape targetShape, HttpBinding binding
+            GoWriter writer, Model model, SymbolProvider symbolProvider,
+            String memberName, Shape targetShape, HttpBinding binding
```
Comment (skmcgrail) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java:998]: nit: Style says that for function parameter definitions that span lines should be formatted such that each parameter is on its own line.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java [hunk 9]
```diff
@@ -936,29 +940,30 @@ private String generateHttpBindingsValue(
...
+                return getCollectionDeserializer(writer, model, symbolProvider, targetValueListShape, binding, operand);
             default:
                 throw new CodegenException("unexpected shape type " + targetShape.getType());
         }
     }
 
     private String getCollectionDeserializer(
-            GoWriter writer, Model model,
+            GoWriter writer, Model model, SymbolProvider symbolProvider,
             Shape targetShape, HttpBinding binding, String operand
```
Comment (skmcgrail) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java:955]: nit: Style says that for function parameter definitions that span lines should be formatted such that each parameter is on its own line.

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java (hunk 7 above)
Comment (jasdel) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java:1011]: cool will update to use that style

File: codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenUtils.java [hunk 10]
```diff
@@ -145,15 +145,53 @@ public static String getSyntheticTypeNamespace() {
...
+            case FLOAT:
+                writer.addUseImports(GoDependency.SMITHY_PTR);
+                return "ptr.Float32(" + operand + ")";
+            case DOUBLE:
+                writer.addUseImports(GoDependency.SMITHY_PTR);
+                return "ptr.Float64(" + operand + ")";
+
+            case TIMESTAMP:
+                writer.addUseImports(GoDependency.SMITHY_PTR);
+                return "ptr.Time(" + operand + ")";
```
Comment (skotambkar) [inline codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenUtils.java:187]: This feels repetitive, as default seems to handle the behavior.

Comment (jasdel) [reply codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/CodegenUtils.java:187]: Yeah its repetitive :/ will see if this can be compressed. There are several cases where we don't want to include the extra dep. and the default case has significantly different behavior.
